---
title: "README"
author: "Daniel DeWaters"
date: "5/7/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# VCF2Ploidy

## Introduction

This package is an update to my original [VCF File Converter tool](https://github.com/dandewaters/VCF-File-Converter), rewritten for R. This R package acts as a companion to the [gbs2ploidy package](https://cran.r-project.org/package=gbs2ploidy). The functions in the GBS2Ploidy package require the input data to be in Heterozygous Allele Depth (HAD) format, however GBS2Ploidy does not include a function for converting from Variant-Calling Format (VCF). The VCF2HAD() function in this package reads in VCF files and returns a data frame in HAD format that can be input to the GBS2Ploidy functions. Alternatively, you can use the VCF2Ploidy() function in this package to achieve this in one step.

## Installation
 
To install, make sure you have the [devtools package](https://cran.r-project.org/package=devtools) installed and loaded. Then run the following commands:

```{r installation}
library(devtools)
install_github("dandewaters/VCF2Ploidy")
library(FARSFunctions)
```

## Vignettes

Read the intro vignette by running install_github with vignettes = TRUE and running the following commands:

```{r vignettes}
library(devtools)
install_github("dandewaters/VCF2Ploidy", build_vignette=TRUE)
vignette("Introduction", package="VCF2Ploidy")
```

## Travis Badge
